probe/kubernetes/client.go

```diff
@@ -0,0 +1,109 @@
+package kubernetes
+
+import (
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type client struct {
+	client *unversioned.Client
+
+	lock      sync.Mutex
+	listeners map[chan<- watch.Event]listener
+}
+
+// TODO: proper config, maybe nil works?
+// TODO: Config should probably be from unversioned.InClusterConfig(), but that
+// needs to run inside a pod.
+func NewClient(api string) (Client, error) {
+	c, err := unversioned.New(&unversioned.Config{Host: api})
+	if err != nil {
+		return nil, err
+	}
+	return &client{
+		client:    c,
+		listeners: map[chan<- watch.Event]listener{},
+	}, nil
+}
+
+// TODO: Subscribe to services too
+func (k *client) AddEventListener(c chan<- watch.Event) error {
+	pods, err := k.client.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), "")
+	if err != nil {
+		return err
+	}
+	services, err := k.client.Services(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), "")
+	if err != nil {
+		pods.Stop()
+		return err
+	}
+	nodes, err := k.client.Nodes().Watch(labels.Everything(), fields.Everything(), "")
+	if err != nil {
+		pods.Stop()
+		services.Stop()
+		return err
+	}
+	k.lock.Lock()
+	defer k.lock.Unlock()
+	k.removeEventListener(c)
+	l := listener{pods: pods, services: services, nodes: nodes}
+	k.listeners[c] = l
+	go l.Listen(c)
+
+	return nil
+}
+
+func (k *client) RemoveEventListener(c chan watch.Event) error {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+	return k.removeEventListener(c)
+}
+
+func (k *client) removeEventListener(c chan<- watch.Event) error {
+	l, ok := k.listeners[c]
+	if ok {
+		delete(k.listeners, c)
+		l.Stop()
+	}
+	return nil
+}
+
+type listener struct {
+	pods, services, nodes watch.Interface
+}
+
+func (l listener) Listen(c chan<- watch.Event) {
+	podsChan := l.pods.ResultChan()
+	servicesChan := l.services.ResultChan()
+	nodesChan := l.nodes.ResultChan()
```

Comment:
Maybe var-block this, and can they just be called pods/services/chan? ([via](https://docs.google.com/document/d/18UTcmy8VoecEaYNb0DrI8NZrI1uJbq-XYDEko0l5Ny4/edit#heading=h.74h5axa7f66g))